placement/cluster_mapping_provider.go

```diff
@@ -0,0 +1,476 @@
+	// then swapping out the pointers atomically
+	mp.RLock()
+	byName := make(map[string]*databaseMappings, len(mp.byName))
+	byRetention := make([]*databaseMappings, len(mp.byRetention))
+	for n, dbm := range mp.byRetention {
+		dbm := dbm.clone()
+		byRetention[n], byName[dbm.name] = dbm, dbm
+	}
+	mp.RUnlock()
+
+	// Update the clone with the new settings
+	for _, db := range p.Databases {
+		if existing := byName[db.Properties.Name]; existing != nil {
+			if err := existing.update(db); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// This is a new mapping - create and apply all rules
+		dbm, err := newDatabaseMappings(db, mp.log)
+		if err != nil {
+			return err
+		}
+
+		byName[db.Properties.Name] = dbm
+		byRetention = append(byRetention, dbm)
+	}
+
+	sort.Sort(databaseMappingsByMaxRetention(byRetention))
+
+	// Swap out the pointers
+	mp.Lock()
+	mp.byRetention, mp.byName = byRetention, byName
+	mp.Unlock()
+	return nil
+}
+
+// watchPlacementChanges is a background goroutine that watches
+// for placement changes
+func (mp *clusterMappingProvider) watchPlacementChanges() {
+	var placement schema.Placement
+	for range mp.watch.C() {
+		val := mp.watch.Get()
+		if err := val.Unmarshal(&placement); err != nil {
+			mp.log.Errorf("could not unmarshal placement data: %v", err)
+			continue
+		}
+
+		mp.log.Infof("received placement version %d", val.Version())
+		mp.update(&placement)
+	}
+	close(mp.closed)
+}
+
+// databaseMappings are the mappings for a given database.
+type databaseMappings struct {
+	name               string
+	version            int32
+	maxRetentionInSecs int32
+	active             []*activeClusterMapping
+	log                xlog.Logger
+}
+
+// newDatabaseMappings returns a newly initialized set of databaseMappings
+func newDatabaseMappings(db *schema.Database, log xlog.Logger) (*databaseMappings, error) {
+	dbm := &databaseMappings{
+		name:               db.Properties.Name,
+		version:            db.Version,
+		maxRetentionInSecs: db.Properties.MaxRetentionInSecs,
+		log:                log,
+	}
+
+	for _, rules := range db.MappingRules {
+		if err := dbm.applyRules(rules); err != nil {
+			return nil, err
+		}
+	}
+
+	return dbm, nil
+}
+
+// clone clones the database mappings
+func (dbm *databaseMappings) clone() *databaseMappings {
+	clone := &databaseMappings{
+		name:               dbm.name,
+		version:            dbm.version,
+		maxRetentionInSecs: dbm.maxRetentionInSecs,
+		log:                dbm.log,
+		active:             make([]*activeClusterMapping, len(dbm.active)),
+	}
+
+	for n, m := range dbm.active {
+		var prior *clusterMapping
+		if m.prior != nil {
+			prior = m.prior.clone()
+		}
+
+		clone.active[n] = &activeClusterMapping{
+			clusterMapping: clusterMapping{
+				cluster:          m.cluster,
+				readCutoverTime:  m.readCutoverTime,
+				writeCutoverTime: m.writeCutoverTime,
+				prior:            prior,
+			},
+			shards: m.shards.Clone(),
+		}
```

Comment:
Might it be best if this logic for cloning a activeClusterMapping live as a method on `activeClusterMapping` so that when fields change on that struct one does not have to remember to update the method on the `databaseMappings` type?